text/0000-array-private-classes

```diff
@@ -0,0 +1,198 @@
+
+This design principle is applied to the other collection classes that expose
+internals too like:
+
+* `List`
+* `Map`
+* persistent `Map`
+* `Vec`
+* persistent `Vec`
+* `Set`
+* `Itertools`
+
+# Motivation
+
+This change brings:
+
+- Applying the design principle of
+[hiding implementation details](https://en.wikipedia.org/wiki/Information_hiding)
+but offer a general and stable interface. Returning interfaces instead of concrete
+classes allows changing the implementation. Usually, one must return the most
+general type that fullfils the contract of the function (in the case of the
+functions discussed in this RFC, iteration).
+- Collections' functions `keys`, `values` and `pairs` definitions are made more
+general. Iterators implementation details are not public. Internal classes used
+by implementation like `*Keys`, `*Values` and `*Pairs` are now
+[opaque data types](https://en.wikipedia.org/wiki/Opaque_data_type). Generally,
+when using these collection classes, clients are not interested by the iterators
+implementation, but by the types these iterators return and that is provided by
+the generic parameters.
+- The generic return signature of these 3 iterating functions is simpler to
+understand for clients of collection classes.
+- Makes the standard library more simple by hiding 18 specialised classes from
+stdlib of which 3 are from `builtin`.
+- The interface `RewindableIterator` is added to create rewindable iterators
+(can be re-start from first value).
+
+This change remains compatible with the existing code base but for client code
+that is directly using the classes `*Keys`, `*Values` and `*Pairs`. A search on
+Github shows that the impact is very limited.
+
+# Detailed design
+
+Iterating functions in collections `keys`, `values` and `pairs` are changed to
+return `Iterator` and the classes that implement these iterators are made private.
+Here are the full implementation of these functions for the `Array` class (changes
+in other collection classes are identical).
+
+```pony
+  fun keys(): Iterator[USize]^ =>
+    """
+    Return an iterator over the indices in the array.
+    """
+    _ArrayKeys[A, this->Array[A]](this)
+
+  fun values(): RewindableIterator[this->A]^ =>
+    """
+    Return an iterator over the values in the array.
+    """
+    _ArrayValues[A, this->Array[A]](this)
+
+  fun pairs(): Iterator[(USize, this->A)]^ =>
+    """
+    Return an iterator over the (index, value) pairs in the array.
+    """
+    _ArrayPairs[A, this->Array[A]](this)
+```
+
+As the function `values` of class `Array` uses an iterator with a `rewind` function
+that is not part of the `Iterator` interface, a new interface `RewindableIterator`
+is added to enable creation of rewindable iterators.
+
+Note: To remain consistent with `Array` behaviour, functions `keys` and `pairs`
+should return a `RewindableIterator` too but we limited the API change to minimum
+as we did not understood why it was not already the case.
```

Comment:
As noted above, I think you should drop this point.

Reply:
I agree that those standard library iterators that *can* be made efficiently rewindable should all be made such.

To me, that's the best definition of "consistent" rather than keeping consistent with the current (artificially limited) behavior.

Reply:
DONE